src/rust/engine/fs/src/snapshot.rs

```diff
@@ -571,6 +603,30 @@ impl StoreFileByDigest<String> for OneOffStoreFileByDigest {
   }
 }
 
+///
+/// A StoreFileByDigest which reads in-memory file contents and writes to a Store, with no caching.
+///
+#[derive(Clone)]
+struct FromMemoryStoreFileByDigest {
+  store: Store,
+  files_content: BTreeMap<File, Bytes>,
+}
+
+impl StoreFileByDigest<String> for FromMemoryStoreFileByDigest {
+  fn store_by_digest(&self, file: File) -> BoxFuture<Digest, String> {
+    let store = self.store.clone();
+    futures::future::done(
+      self
+        .files_content
+        .get(&file)
```

Comment:
Almost commented that this is unfortunate, then noticed that it is `Bytes`, huzzah.